kie-issues#85: Add more Git operations on KIE Sandbox

jstastny-cz opened the pull request:
Resolves kiegroup/kie-issues#85 .

New:
* FileSwitcher menuItems contain git status indicator (Modified, Deleted, Added) and a hoverable icon for reverting the change. On hover a popover appears with inline alert contents.
* New Component `GitStatusIndicatorActions.tsx` implements two variants of how actions are presented: popover vs. dropdown based design, details hidden under the component itself. It's passed to GitStatusIndicator.tsx as children.
* WorkspaceStatusIndicator contains statuses for all files in workspace, e.g. `M | A | D` in case there are files in such states.
* WorkspaceStatusIndicator contains a dropdown variant of the git actions.
* Implemented revert operation as checkout of files from HEAD using isomorphic-git in `GitService.ts`. Invoked through the GitStatusIndicatorActions.tsx passing
  * single file in FileSwitcher menu item
  * all modified files in WorkspaceStatusIndicator dropdown
* Introduced event type `WS_CHECKOUT_HEAD_FILES` to signal changes to the workspace itself. The message contains relativePaths to all the files that were processed - allows for checking if update of current Editor contents should be refreshed.

Noteworthy refactoring:
* FileSwitcher now shows also the active file, to be able to display git operations related just to the current file.
* FilesMenuItems component reused in HomePage.tsx - reduction of duplicated code and unification - it was not possible to show the files as SVG cards on HomePage before.
* `WorkspaceHooks.tsx` signature changed to `useWorkspaceGitStatusPromise(workspace: WorkspaceDescriptor | undefined)` to enable reusing in HomePage.tsx
* Workspace toolbar is shown also for Local WorkspaceKind now, it enables the additional git actions also for Local workspaces.

paulovmr commented:
@jstastny-cz This PR should fix the CI error: https://github.com/kiegroup/kie-tools/pull/1484

jstastny-cz commented:
@caponetto if you could please review too, there are changes in workspaces-git-fs affecting serverless-logic-web-tools too:
1. WorkspaceHooks.tsx signature changed to `useWorkspaceGitStatusPromise(workspace: WorkspaceDescriptor | undefined)`
2. WorkspaceDescriptorService.ts removing broadcast in `bumpLastUpdatedDate` function which was causing duplicate git status query on each change in the editor. Originally this was added as part of the fix for refreshing workspaces order on the HomePage when a workspace is updated async, like in a different browser window. This will be reported as an issue and needs a different fix.

jstastny-cz commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -478,26 +483,9 @@ export function NewModelCard(props: { title: string; extension: SupportedFileExt
-    return [...models, ...otherFiles];
-  }, [models, otherFiles]);
+  const workspaceGitStatusPromise = useWorkspaceGitStatusPromise(workspacePromise.data?.descriptor);
+  const [filesDropdownMode, setFilesDropdownMode] = useState(FilesDropdownMode.LIST_MODELS_AND_OTHERS);
```
Current implementation of the dropdown mode is causing issues in home page with width, investigating if View others/View models can be implemented as a drilldown instead of Split as it is now.

tiagobento commented on packages/online-editor/src/workspace/components/GitStatusIndicatorActions.tsx:
```diff
@@ -0,0 +1,513 @@
+  InfoCircleIcon,
+  UndoAltIcon,
+  WarningTriangleIcon,
+} from "@patternfly/react-icons";
```
Wrong imports... please always prefer the most specific ones. Also, double-check other places on the PR.

tiagobento commented on packages/online-editor/src/home/HomePage.tsx:
```diff
@@ -558,24 +545,17 @@ export function WorkspacesListDrawerPanelContent(props: { workspaceId: string |
+              workspace={workspace}
+              onSelectFile={() => {}}
+              workspaceGitStatusPromise={workspaceGitStatusPromise}
+              workspaces={workspaces}
```
This is not good.. you don't need to pass `workspaces` as props.. you should be able to `useWorkspaces` where you need it.

tiagobento replied:
There's a lot of other places with this same behavior too.

jstastny-cz replied:
Right, this is what I was asking earlier - if we should be passing rather than using the use* hooks or not. Is there a rule of thumb which are better to pass and which are safe to invoke?
The concern was not to re-evaluate those like workspaceGitStatus, which might be expensive.

tiagobento replied:
Hmmmm, yeah, I can see how that's confusing.. Usually, as hooks should be done in a way that it's safe to call them multiple times, but the git status hook is an exception I guess... we could've materialized it to a Context rather then a simple hook.. this would make it safe for any component to call it without performance implications.. Let's say that for all hooks it's okay to call them any amount of times, except for this one for now.. :)

tiagobento commented on packages/online-editor/src/i18n/locales/en.ts:
```diff
@@ -626,4 +626,16 @@ export const en: OnlineI18n = {
       organizations: "GitHub organizations",
     },
   },
+  gitStatusIndicatorActions: {
+    revert: {
+      title: "Revert Changes",
+      description: ["This action is permanent. Following file will be reverted to the last commit:", wrapped("paths")],
+      confirmButtonText: "Revert Permanently",
+    },
+    revertAll: {
+      title: "Revert All Changes",
+      description: ["This action is permanent. Following files will be reverted to the last commit:", wrapped("paths")],
+      confirmButtonText: "Revert Permanently",
+    },
+  },
```
At some point, non-programmers will need to write i18n files. Our setup is already more complicated than .properties files, so we should also avoid complex constructs here.. IDK if there's any other place using an array, but if possible, we should avoid doing so...

jstastny-cz replied:
That's for the I18nWrapper component that is used also elsewhere. Just grep `wrapped` in this file. Does that mean we should not use that construct?

tiagobento replied:
Oh, alright. If that was already in place, then it's okay. I don't exactly remember when we did this, but if the array is necessary, then let's leave it be. We'll have to revisit it at some point when we go back to working on i18n again.

jstastny-cz replied:
I've changed to plain string instead, there was no actual value brought by this construct.

tiagobento commented on packages/online-editor/src/workspace/components/GitStatusIndicatorActions.tsx:
```diff
@@ -0,0 +1,513 @@
+          relativePath: props.currentWorkspaceFile.relativePath,
+        }))
+      ) {
+        props.onDeletedWorkspaceFile && props.onDeletedWorkspaceFile();
```
```suggestion
        props.onDeletedWorkspaceFile?.();
```

tiagobento commented on packages/online-editor/src/workspace/components/GitStatusIndicatorActions.tsx:
```diff
@@ -0,0 +1,513 @@
+
+  const doRevert = React.useCallback(
+    async (filepaths: string[]) => {
+      await props.workspaces.checkoutFilesFromLocalHead({
+        workspaceId: props.workspaceDescriptor.workspaceId,
+        ref: props.workspaceDescriptor.origin.branch,
+        filepaths,
+      });
+      if (
+        props.currentWorkspaceFile &&
+        !(await props.workspaces.existsFile({
+          workspaceId: props.workspaceDescriptor.workspaceId,
+          relativePath: props.currentWorkspaceFile.relativePath,
+        }))
+      ) {
```
There are many places passing `workspaces` down as a prop just for this moment here to happen.. That is not very good IMHO.

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+  }, [props.workspaceDescriptor, props.workspaceGitStatusPromise]);
+
+  const isThereDeletedFile = React.useMemo(() => {
+    const maybeWorkspaceFile = props.workspaceFile;
```
No reason for this variable to exist, honestly..

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+  const isThereDeletedFile = React.useMemo(() => {
+    const maybeWorkspaceFile = props.workspaceFile;
+    return (
+      (!maybeWorkspaceFile && deletedFilesForWorkspace.length) ||
+      (maybeWorkspaceFile &&
+        deletedFilesForWorkspace.some((deleted) => deleted.relativePath === maybeWorkspaceFile.relativePath))
```
Why the two possibilities? They seem equivalent to me.

jstastny-cz replied:
So we have an optional workspaceFile property passed to the indicator - if that's set it means the indicator applies just to that single file (in file menu).
When it's not set, it means the indicator is for whole workspace.

So when we don't have a specific file it is enough to check length of the deletedFilesForWorkspace array. But if we have a specific WorkspaceFile set, we need to check specifically, if deletedFilesForWorkspace contain exactly same relativePath.

jstastny-cz replied:
Anyway this method is not needed, removed now.

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+      workspaceGitStatus: props.workspaceGitStatusPromise?.data ?? prev,
+      file: props.workspaceFile,
+    });
+    if (props.onStatusChanged) {
+      props.onStatusChanged(
+        switchExpression(status, {
+          pending: isThereDeletedFile ? "deleted" : "modified",
+          synced: "synced",
+          default: undefined,
+        })
+      );
+    }
```
```suggestion
      props.onStatusChanged?.(
        switchExpression(status, {
          pending: isThereDeletedFile ? "deleted" : "modified",
          synced: "synced",
          default: undefined,
        })
      );
```

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+      tooltips.push(
+        tooltipForStageStatus(
+          props.workspaceGitStatusPromise?.data.fileStageStatuses.find(
+            (it) => it.path === maybeWorkspaceFile.relativePath
```
If you don't know what to call a variable, you can resort to object destruction... 
```suggestion
            ({path}) => path === maybeWorkspaceFile.relativePath
```

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+        )
+      );
+    } else {
+      new Set(props.workspaceGitStatusPromise?.data.fileStageStatuses.map((it) => it.status)).forEach((it) => {
+        tooltips.push(tooltipForStageStatus(it));
+      });
```
Same here.

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+        deleted: deletedTooltip,
+        default: <></>,
+      });
+    const maybeWorkspaceFile = props.workspaceFile;
+    if (maybeWorkspaceFile) {
```
Really not necessary for this variable to exist.

jstastny-cz replied:
if I put `if(props.workspaceFile)` then it does not serve as type guard when accessing in the find function below. Am I missing something, or is it fine to put `props.workspaceFile?.relativePath` below, even though I have checked it's value in this if statement?

tiagobento replied:
Hmmm.. interesting.. Looks like a limitation from the TypeScript type inference.. I'd rather use an extra `?` then naming an unnecessary temporary variable.. WDYT?

thiagoelg replied:
I've had this problem before, not sure why it happens 😕 
Destructuring the props outside the `useMemo` and then checking it might be a solution... not sure if you guys would like it though.

```js
const { workspaceFile } = props;

...

const indicatorTooltips = React.useMemo(() => {
  ...
  if (workspaceFile) {
     ...
   }
   ...
}, [workspaceFile]);
```

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+  workspaceDescriptor: WorkspaceDescriptor;
+  workspaces: WorkspacesContextType;
+  workspaceGitStatusPromise: PromiseState<WorkspaceGitStatusType>;
+  currentWorskpaceFile: WorkspaceFile;
```
Typo.

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
-export function WorkspaceStatusIndicator(props: { workspace: ActiveWorkspace }) {
+export function WorkspaceStatusIndicator(props: {
+  workspaceDescriptor: WorkspaceDescriptor;
+  workspaces: WorkspacesContextType;
```
Not good to have this as a prop, as this is a context.

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -96,46 +249,92 @@ export function WorkspaceStatusIndicator(props: { workspace: ActiveWorkspace })
+        workspaces={props.workspaces}
+        currentWorkspaceFile={props.currentWorskpaceFile}
+        onDeletedWorkspaceFile={props.onDeletedWorkspaceFile}
+        workspaceGitStatus={props.workspaceGitStatusPromise}
```
Please keep prop names consistent.. `workspaceGitStatus` -> `workspaceGitStatusPromise`

tiagobento commented on packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx:
```diff
@@ -96,46 +249,92 @@ export function WorkspaceStatusIndicator(props: { workspace: ActiveWorkspace })
+  if (args.workspaceGitStatus === undefined) {
+    return WorkspaceGitLocalChangesStatus.unknown;
+  }
+  const maybeFile = args.file;
+  if (maybeFile) {
```
:D?

tiagobento reviewed (changes requested):
Great job! Left some comments inline, and as a general comment, I'd say let's try to keep new concepts cohesive. I know naming things is really hard, but it helps if the names are consistent. For example:

WS_CHECKOUT_HEAD_FILES
checkoutFilesFromLocalHead
checkoutLocalHead

They are part of the same concept, but they have slight variations on their names, making it hard to know for sure the they are really related or not.

I'll keep reviewing and will test it locally, but this first round is ready.

tiagobento commented on packages/online-editor/src/workspace/components/GitStatusIndicatorActions.tsx:
```diff
@@ -0,0 +1,513 @@
+
+export type GitStatusIndicatorActionsType = GitStatusIndicatorActionsDropdown | GitStatusIndicatorActionsPopover;
+
+export const GitStatusIndicatorActions = (props: GitStatusIndicatorActionsType) => {
```
It's weird that a props type is not called `Props` at the end...

jstastny-cz commented:
Latest changes include not only resolution of previous comments, but also further refactoring of FileSwitcher.tsx. The problem there was that the initial refactoring didn't account for displaying otherFiles inside a Split, which doubled width of the file menu. That was alright in EditorPage, but was causing issues in HomePage where I reuse the same component from FileSwitcher. To achieve consistent rendering in both places, I've switched to display otherFiles as another level of drilldownMenu, which keeps the width same all the time.
Next issue on HomePage were nested scrolls when there were too many files, while in FileSwitcher we need to put limits on max height, in HomePage this resulted in nested scrollbars, making interaction problematic. So I've introduced a conditional `isNoHeightMaxLimit` which conditionally removes the restrictions for file listings. One of the implications is that `View other files` menu item in HomePage is displayed after all the models (all the way down), while in EditorPage the button stays accessible all the time - because of the restricted height of preceding files listing.

caponetto reviewed (commented):
Overall looks good to me. Thanks, @jstastny-cz!
I've left some minor comments inline, please check if they make sense. :)

jstastny-cz commented on packages/workspaces-git-fs/src/services/GitService.tsx:
```diff
@@ -158,6 +168,22 @@ export class GitService {
     });
   }
 
+  public async checkoutFilesFromLocalHead(args: {
```
@tiagobento what do you think? We have discusses this before and thought separate methods would be better - though I am not sure now if it meant already on GitService or only at the WorkspacesWorkerApi level.

tiagobento replied:
I think we could rename the first one to `checkoutAll` or `checkoutEverything`.. I don't think we should be exposing the full power of the `checkout` method from `isomorphic-git`.

jstastny-cz commented on packages/workspaces-git-fs/src/services/GitService.tsx:
```diff
@@ -68,6 +68,16 @@ export interface RemoteRefArgs {
+  modified = "modified",
+  deleted = "deleted",
+}
+export type FileStageStatusType = {
```
I agree, I've switched now to following:
```
export enum FileModificationStatus {
  added = "added",
  modified = "modified",
  deleted = "deleted",
}
export type UnstagedModifiedFilesStatusEntryType = {
  path: string;
  status: FileModificationStatus;
};
```
If you have a better proposal, please, help.

jstastny-cz commented on packages/workspaces-git-fs/src/context/WorkspacesContext.tsx:
```diff
@@ -148,6 +151,8 @@ export interface WorkspacesContextType {
 
   hasLocalChanges(args: { workspaceId: string }): Promise<boolean>;
 
+  getUnstagedModifiedFileRelativePaths(args: { workspaceId: string }): Promise<FileStageStatusType[]>;
```
Renamed api methods to:
* `unstagedModifiedFilesStatus`
* `kieSandboxWorkspacesGit_getUnstagedModifiedFilesStatus`

Variables holding the value are now named `unstagedModifiedFilesStatus`. And while iterating through contents, the local var is named `stageStatusEntry`.

jstastny-cz commented:
I will shortly push new changes. Due to limited space on HomePage, it was not convenient to reuse same menu as in FileSwitcher, thus reverting those modifications.

jstastny-cz commented on packages/online-editor/src/editor/Toolbar/FileSwitcher.tsx:
```diff
@@ -371,20 +377,26 @@ export function FileSwitcher(props: { workspace: ActiveWorkspace; workspaceFile:
                     Current
                   </MenuItem>
-                  {workspacesMenuItems}
+                  <MenuGroup
```
@tiagobento The purpose of the MenuGroup maxHeight here was only to enable scrolling inside MenuContent while keeping the `Current >` menu item all the time at the top of the menu. In the last commit the overflow was moved to MenuContent itself thus scrolling all items including the `Current >`, I wanted to check if that was on purpose, if so, I will remove the maxHeight specification here.

jstastny-cz commented:
Scope was limited to the Editor page only, home page changes were reverted.

thiagoelg commented on packages/online-editor/src/filesList/FileDataList.tsx:
```diff
@@ -42,31 +46,74 @@ export function getFileDataListHeight(file: WorkspaceFile) {
+            distance={5}
+            position={"top-start"}
+            content={
+              props.file.nameWithoutExtension +
```
Shouldn't we follow the same rule as before here?
`const fileName = props.isEditable ? props.file.nameWithoutExtension : props.file.name;`

Showing the file extension if it's not editable.
